java-checks/src/main/java/org/sonar/java/checks/ReturnInFinallyCheck.java

```diff
@@ -52,53 +58,131 @@ public void visitTryStatement(TryStatementTree tree) {
     scan(tree.resources());
     scan(tree.block());
     scan(tree.catches());
-    if (tree.finallyBlock() != null) {
-      isInFinally.push(true);
-      scan(tree.finallyBlock());
+    BlockTree finallyBlock = tree.finallyBlock();
+    if (finallyBlock != null) {
+      isInFinally.push(finallyBlock.kind());
+      scan(finallyBlock);
       isInFinally.pop();
     }
   }
 
   @Override
   public void visitMethod(MethodTree tree) {
-    isInFinally.push(false);
+    isInFinally.push(tree.kind());
     super.visitMethod(tree);
     isInFinally.pop();
   }
 
   @Override
+  public void visitForStatement(ForStatementTree tree) {
+    isInFinally.push(tree.kind());
+    super.visitForStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
+  public void visitForEachStatement(ForEachStatement tree) {
+    isInFinally.push(tree.kind());
+    super.visitForEachStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
+  public void visitWhileStatement(WhileStatementTree tree) {
+    isInFinally.push(tree.kind());
+    super.visitWhileStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
+  public void visitDoWhileStatement(DoWhileStatementTree tree) {
+    isInFinally.push(tree.kind());
+    super.visitDoWhileStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
+  public void visitSwitchStatement(SwitchStatementTree tree) {
+    isInFinally.push(tree.kind());
+    super.visitSwitchStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
   public void visitReturnStatement(ReturnStatementTree tree) {
-    reportIssue(tree.returnKeyword(), "return");
+    reportIssue(tree.returnKeyword(), tree.kind());
```

Comment:
All your first parameters are `SyntaxToken`, just modify the method to have a `SyntaxToken` as first parameter instead of a tree. It would allow you to directly access it's value, instead of using that useless `statement()` method